RHDEVDOCS-5509: Configuring build strategies (#63401)
openshift/openshift-docs: modules/ob-strategies-with-different-resources.adoc

// This module is included in the following assembly:
//
// builds/configuring-build-strategies.adoc

:_content-type: CONCEPT
[id="ob-strategies-with-different-resources_{context}"]
= Strategies with different resources

As a strategy administrator, you can define multiple types of the same strategy with varying limits on the resources. The following example uses the same Kaniko strategy with small and medium limits defined for the resources:

.Example: `kaniko` strategy with small limit 
[source,yaml]
----
apiVersion: shipwright.io/v1beta1
kind: ClusterBuildStrategy
metadata:
  name: kaniko-small
spec:

>   buildSteps:

Reviewer: ```suggestion
  steps:
```

    - name: build-and-push
      image: gcr.io/kaniko-project/executor:v1.8.1
      workingDir: $(params.shp-source-root)
      securityContext:
        runAsUser: 0
        capabilities:
          add:
            - CHOWN
            - DAC_OVERRIDE
            - FOWNER
            - SETGID
            - SETUID
            - SETFCAP
            - KILL
      env:
        - name: CONTAINER_CONFIG
          value: /tekton/home/.container
        - name: AWS_ACCESS_KEY_ID
          value: NOT_SET
        - name: AWS_SECRET_KEY
          value: NOT_SET
      command:
        - /kaniko/executor
      args:
        - --skip-tls-verify=true
        - --containerfile=$(build.containerfile)
        - --context=$(params.shp-source-context)
        - --destination=$(params.shp-output-image)
        - --snapshotMode=redo
        - --push-retry=3
      resources:
        limits:
          cpu: 250m
          memory: 65Mi
        requests:
          cpu: 250m
          memory: 65Mi
----

.Example: `kaniko` strategy with medium limit 
[source,yaml]
----

> apiVersion: shipwright.io/v1alpha1

Reviewer: ```suggestion
apiVersion: shipwright.io/v1beta1
```

kind: ClusterBuildStrategy
metadata:
  name: kaniko-medium
spec:

>   buildSteps:

Reviewer: ```suggestion
  steps:
```

        capabilities:
          add:
            - CHOWN
            - DAC_OVERRIDE
            - FOWNER
            - SETGID
            - SETUID
            - SETFCAP
            - KILL
      env:
        - name: CONTAINER_CONFIG
          value: /tekton/home/.container
        - name: AWS_ACCESS_KEY_ID
          value: NOT_SET
        - name: AWS_SECRET_KEY
          value: NOT_SET
      command:
        - /kaniko/executor
      args:
        - --skip-tls-verify=true
        - --containerfile=$(build.containerfile)
        - --context=$(params.shp-source-context)
        - --destination=$(params.shp-output-image)
        - --snapshotMode=redo
        - --push-retry=3
      resources:
        limits:
          cpu: 500m
          memory: 1Gi
        requests:
          cpu: 500m
          memory: 1Gi
----

The above example provides a strategy administrator more control over the steps resource definition. When the resource definition is configured for a strategy, you must reference the strategy in your `Build` CR, as shown in the following example: 

[source,yaml]
----
apiVersion: shipwright.io/v1beta1
kind: Build
metadata:
  name: kaniko-medium
spec:
  source:

>     url: https://github.com/shipwright-io/sample-go

Reviewer: ```suggestion
    git:
      url: https://github.com/shipwright-io/sample-go
```

          memory: 65Mi
----

.Example: `kaniko` strategy with medium limit 
[source,yaml]
----
apiVersion: shipwright.io/v1beta1
kind: ClusterBuildStrategy
metadata:
  name: kaniko-medium
spec:
  steps:
    - name: build-and-push
      image: gcr.io/kaniko-project/executor:v1.8.1
      workingDir: $(params.shp-source-root)
      securityContext:
        runAsUser: 0
        capabilities:
          add:
            - CHOWN
            - DAC_OVERRIDE
            - FOWNER
            - SETGID
            - SETUID
            - SETFCAP
            - KILL
      env:
        - name: CONTAINER_CONFIG
          value: /tekton/home/.container
        - name: AWS_ACCESS_KEY_ID
          value: NOT_SET
        - name: AWS_SECRET_KEY
          value: NOT_SET
      command:
        - /kaniko/executor
      args:
        - --skip-tls-verify=true
        - --containerfile=$(build.containerfile)
        - --context=$(params.shp-source-context)
        - --destination=$(params.shp-output-image)
        - --snapshotMode=redo
        - --push-retry=3
      resources:
        limits:
          cpu: 500m
          memory: 1Gi
        requests:
          cpu: 500m
          memory: 1Gi
----

> The above example provides a strategy administrator more control over the steps resource definition. When the resource definition is configured for a strategy, you must reference the strategy in your `Build` CR, as shown in the following example: 

Author: The previous two examples are the combination of the same aspect, that is, providing more flexibility in defining a  resource limits for a build strategy.

// This module is included in the following assembly:
//
// builds/configuring-build-strategies.adoc

:_content-type: CONCEPT
[id="ob-strategies-with-different-resources_{context}"]
= Strategies with different resources

As a strategy administrator, you can define multiple types of the same strategy with varying limits on the resources. The following examples use the same `kaniko` strategy with small and medium limits defined for the resources:

> .Example: `kaniko` strategy with small limit 

Reviewer 2: You might want to label this "Example "A" since another follows: 

```suggestion
.Example A: `kaniko` strategy with small limit 
```

// This module is included in the following assembly:
//
// builds/configuring-build-strategies.adoc

:_content-type: CONCEPT
[id="ob-strategies-with-different-resources_{context}"]
= Strategies with different resources

> As a strategy administrator, you can define multiple types of the same strategy with varying limits on the resources. The following examples use the same `kaniko` strategy with small and medium limits defined for the resources.

Reviewer 3: https://stylepedia.net/style/6.1/#redundant

These examples provide a strategy administrator more control over the steps resource definition. 

[id="example-kaniko-small-limit"]
== Kaniko strategy with small limit

> As a strategy administrator, you can define the `spec.steps.resource` field with small resource limit for the `kaniko` strategy, as shown in the following example:

Reviewer 3: https://stylepedia.net/style/6.1/#redundant

[source,yaml]
----
apiVersion: shipwright.io/v1beta1
kind: ClusterBuildStrategy
metadata:
  name: kaniko-small
spec:
  steps:
    - name: build-and-push
      image: gcr.io/kaniko-project/executor:v1.8.1
      workingDir: $(params.shp-source-root)
      securityContext:
        runAsUser: 0
        capabilities:
          add:
            - CHOWN
            - DAC_OVERRIDE
            - FOWNER
            - SETGID
            - SETUID
            - SETFCAP
            - KILL
      env:
        - name: CONTAINER_CONFIG
          value: /tekton/home/.container
        - name: AWS_ACCESS_KEY_ID
          value: NOT_SET
        - name: AWS_SECRET_KEY
          value: NOT_SET
      command:
        - /kaniko/executor
      args:
        - --skip-tls-verify=true
        - --containerfile=$(build.containerfile)
        - --context=$(params.shp-source-context)
        - --destination=$(params.shp-output-image)
        - --snapshotMode=redo
        - --push-retry=3
      resources:
        limits:
          cpu: 250m
          memory: 65Mi
        requests:
          cpu: 250m
          memory: 65Mi
----

[id="example-kaniko-medium-limit"]
== Kaniko strategy with medium limit

> As a strategy administrator, you can define the `spec.steps.resource` field with medium resource limit for the `kaniko` strategy, as shown in the following example:

Reviewer 3: https://stylepedia.net/style/6.1/#redundant

.Example: `kaniko` strategy with medium limit 
[source,yaml]
----

> apiVersion: shipwright.io/v1beta1
> kind: ClusterBuildStrategy
> metadata:
>   name: kaniko-medium
> spec:
>   steps:
>     - name: build-and-push
>       image: gcr.io/kaniko-project/executor:v1.8.1
>       workingDir: $(params.shp-source-root)
>       securityContext:
>         runAsUser: 0
>         capabilities:
>           add:
>             - CHOWN
>             - DAC_OVERRIDE
>             - FOWNER
>             - SETGID
>             - SETUID
>             - SETFCAP
>             - KILL
>       env:
>         - name: CONTAINER_CONFIG
>           value: /tekton/home/.container
>         - name: AWS_ACCESS_KEY_ID
>           value: NOT_SET
>         - name: AWS_SECRET_KEY
>           value: NOT_SET
>       command:
>         - /kaniko/executor
>       args:
>         - --skip-tls-verify=true
>         - --containerfile=$(build.containerfile)
>         - --context=$(params.shp-source-context)
>         - --destination=$(params.shp-output-image)
>         - --snapshotMode=redo
>         - --push-retry=3
>       resources:
>         limits:
>           cpu: 500m
>           memory: 1Gi
>         requests:
>           cpu: 500m
>           memory: 1Gi

Reviewer: ```suggestion
---
apiVersion: shipwright.io/v1beta1
kind: ClusterBuildStrategy
metadata:
  name: buildah-medium
spec:
  steps:
    - name: build-and-push
      image: quay.io/containers/buildah:v1.31.0
      workingDir: $(params.shp-source-root)
      securityContext:
        capabilities:
          add:
          - "SETFCAP"
      command:
        - /bin/bash
      args:
        - -c
        - |
          set -euo pipefail

          # Parse parameters
        # ...
        # That's the separator between the shell script and its args
        - --
        - --context
        - $(params.shp-source-context)
        - --dockerfile
        - $(build.dockerfile)
        - --image
        - $(params.shp-output-image)
        - --build-args
        - $(params.build-args[*])
        - --registries-block
        - $(params.registries-block[*])
        - --registries-insecure
        - $(params.registries-insecure[*])
        - --registries-search
        - $(params.registries-search[*])
      resources:
        limits:
          cpu: 500m
          memory: 1Gi
        requests:
          cpu: 500m
          memory: 1Gi
  parameters:
    - name: build-args
      description: "The values for the args in the Dockerfile. Values must be in the format KEY=VALUE."
      type: array
      defaults: []
    # ...
```

----

When the resource definition is configured for a strategy, you must reference the strategy in your `Build` CR, as shown in the following example: 

[source,yaml]
----

> apiVersion: shipwright.io/v1beta1
> kind: Build
> metadata:
>   name: kaniko-medium
> spec:
>   source:
>     git:
>       url: https://github.com/shipwright-io/sample-go
>     contextDir: container-build
>   strategy:
>     name: kaniko
>     kind: ClusterBuildStrategy
>   containerfile: Containerfile

Reviewer: ```suggestion
apiVersion: shipwright.io/v1beta1
kind: Build
metadata:
  name: buildah-medium
spec:
  source:
    git:
      url: https://github.com/shipwright-io/sample-go
    contextDir: container-build
  strategy:
    name: buildah-medium
    kind: ClusterBuildStrategy
  # ...
```

// This module is included in the following assembly:
//
// * builds/configuring-build-strategies.adoc

:_content-type: REFERENCE
[id="ob-strategies-with-different-resources_{context}"]
= Strategies with different resources

Define multiple types of the same strategy with varying limits on the resources. The following examples use the same `buildah` strategy with small and medium limits defined for the resources.

> These examples provide a strategy administrator more control over the steps resource definition.

Reviewer 4: Double check "steps."

> [id="example-buildah-small-limit"]

Reviewer 4: ```suggestion
[id="example-buildah-small-limit_{context}"]
```

== Buildah strategy with small limit

> Define the `spec.steps[].resources` field with small resource limit for the `buildah` strategy, as shown in the following example:

Reviewer 4: ```suggestion
Define the `spec.steps[].resources` field with a small resource limit for the `buildah` strategy, as shown in the following example:
```

    - name: build-and-push
      image: quay.io/containers/buildah:v1.31.0
      workingDir: $(params.shp-source-root)
      securityContext:
        capabilities:
          add:
          - "SETFCAP"
      command:
        - /bin/bash
      args:
        - -c
        - |
          set -euo pipefail
          # Parse parameters
        # ...
        # That's the separator between the shell script and its args
        - --
        - --context
        - $(params.shp-source-context)
        - --dockerfile
        - $(build.dockerfile)
        - --image
        - $(params.shp-output-image)
        - --build-args
        - $(params.build-args[*])
        - --registries-block
        - $(params.registries-block[*])
        - --registries-insecure
        - $(params.registries-insecure[*])
        - --registries-search
        - $(params.registries-search[*])
      resources:
        limits:
          cpu: 250m
          memory: 65Mi
        requests:
          cpu: 250m
          memory: 65Mi
  parameters:
    - name: build-args
      description: "The values for the args in the Dockerfile. Values must be in the format KEY=VALUE."
      type: array
      defaults: []
    # ...
----

> [id="example-buildah-medium-limit"]

Reviewer 4: ```suggestion
[id="example-buildah-medium-limit_{context}"]
```

== Buildah strategy with medium limit

> Define the `spec.steps[].resources` field with medium resource limit for the `buildah` strategy, as shown in the following example:

Reviewer 4: ```suggestion
Define the `spec.steps[].resources` field with a medium resource limit for the `buildah` strategy, as shown in the following example:
```

    - name: build-and-push
      image: quay.io/containers/buildah:v1.31.0
      workingDir: $(params.shp-source-root)
      securityContext:
        capabilities:
          add:
          - "SETFCAP"
      command:
        - /bin/bash
      args:
        - -c
        - |
          set -euo pipefail
          # Parse parameters
        # ...
        # That's the separator between the shell script and its args
        - --
        - --context
        - $(params.shp-source-context)
        - --dockerfile
        - $(build.dockerfile)
        - --image
        - $(params.shp-output-image)
        - --build-args
        - $(params.build-args[*])
        - --registries-block
        - $(params.registries-block[*])
        - --registries-insecure
        - $(params.registries-insecure[*])
        - --registries-search
        - $(params.registries-search[*])
      resources:
        limits:
          cpu: 500m
          memory: 1Gi
        requests:
          cpu: 500m
          memory: 1Gi
  parameters:
    - name: build-args
      description: "The values for the args in the Dockerfile. Values must be in the format KEY=VALUE."
      type: array
      defaults: []
    # ...
----

> When the resource definition is configured for a strategy, you must reference the strategy in your `Build` CR, as shown in the following example:

Reviewer 4: "is configured for" or "includes"?

Author: We cannot use "includes" because a resource definition does not include a startegy, it is actually the other way round (a strategy includes a resource definition). I have tweaked the sentence a bit for clarity:
After configuring the resource definition for a strategy, you must reference the strategy in your `Build` CR
